testing/scenario_app/ios/Scenarios/ScenariosTests/AppLifecycleTests.m

```diff
@@ -0,0 +1,144 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#import <Flutter/Flutter.h>
+#import <XCTest/XCTest.h>
+#import "../Scenarios/ScreenBeforeFlutter.h"
+
+@interface AppLifecycleTests : XCTestCase
+@end
+
+@implementation AppLifecycleTests
+
+- (void)setUp {
+  [super setUp];
+  self.continueAfterFailure = NO;
+}
+
+- (void)testLifecycleChannel {
+  XCTestExpectation* engineStartedExpectation = [self expectationWithDescription:@"Engine started"];
+
+  // Let the engine finish booting (at the end of which the channels are properly set-up) before
+  // moving onto the next step of showing the next view controller.
+  ScreenBeforeFlutter* rootVC = [[ScreenBeforeFlutter alloc] initWithEngineRunCompletion:^void() {
+    [engineStartedExpectation fulfill];
+  }];
+
+  [self waitForExpectations:[NSArray arrayWithObject:engineStartedExpectation] timeout:5];
+
+  UIApplication* application = UIApplication.sharedApplication;
+  application.delegate.window.rootViewController = rootVC;
+  FlutterEngine* engine = rootVC.engine;
+
+  NSMutableArray* lifecycleExpectations = [NSMutableArray arrayWithCapacity:10];
+  NSMutableArray* lifecycleEvents = [NSMutableArray arrayWithCapacity:10];
+
+  [lifecycleExpectations addObject:[[XCTestExpectation alloc]
+                                       initWithDescription:@"A loading FlutterViewController goes "
+                                                           @"through AppLifecycleState.inactive"]];
+  [lifecycleExpectations
+      addObject:[[XCTestExpectation alloc]
+                    initWithDescription:
+                        @"A loading FlutterViewController goes through AppLifecycleState.resumed"]];
+
+  FlutterViewController* flutterVC = [rootVC showFlutter];
+  [engine.lifecycleChannel setMessageHandler:^(id message, FlutterReply callback) {
+    if (lifecycleExpectations.count == 0) {
+      XCTFail(@"Unexpected lifecycle transition: %@", message);
+    }
+    [lifecycleEvents addObject:message];
+    [[lifecycleExpectations objectAtIndex:0] fulfill];
+    [lifecycleExpectations removeObjectAtIndex:0];
+  }];
+
+  [self waitForExpectations:lifecycleExpectations timeout:5];
+
+  // Expected sequence from showing the FlutterViewController is inactive and resumed.
+  BOOL lifecyclesAsExpected = [lifecycleEvents
```

Comment:
Instead of doing this array checking, you can enforce expectation order.  Something like:
```objective-c
XCTestExpectation *inactiveExpectation = [self expectationWithDescription:@"inactive"];
XCTestExpectation *resumedExpectation = [self expectationWithDescription:@"resume"];
[engine.lifecycleChannel setMessageHandler:^(id message, FlutterReply callback) {
  if ([message isEqualToString:@"AppLifecycleState.inactive"]) {
    [inactiveExpectation fulfill];
  else if ([message isEqualToString:@"AppLifecycleState.resumed"]) {
    [resumedExpectation fulfill];
  } else { 
    XCTFail(@"Unexpected lifecycle transition: %@", message);
  }
  }];
[self waitForExpectations:@[inactiveExpectation, resumedExpectation] timeout:5.0 enforceOrder:YES];
```

Reply:
What will the message be in the next PR that you're going to forbid that isn't covered by the `XCTFail` fall-through?  Can you [invert the expectation](https://developer.apple.com/documentation/xctest/xctestexpectation/2806573-inverted?language=objc)?  I can write some example code with a concrete example.

Reply:
Oh, that's really useful. I think I'll go with your approach when I fix https://github.com/flutter/flutter/issues/37226 and flip the check around. 

How would I synchronize at each step to make sure nothing happens then? Would it still be timeout based? 

Otherwise, I would expect the error message to be in the shape of:

Expecting transitions inactive, resume. But you got inactive, paused, inactive, resume, inactive, resume.

Reply:
-inverted is timeout based liked all the expectations, unfortunately, and enforcing order doesn't make sense.  So it will fail immediately if fulfilled, but then will need to wait for the timeout to succeed
```
[self waitForExpectations:@[inactiveExpectation, resumedExpectation] timeout:5.0 enforceOrder:YES];
[self waitForExpectations:@[invertedExpectation] timeout:1.0]; // inactiveExpectation, resumedExpectation already fulfilled so wait less time?
```
However, what I wrote above should fail with "inactive, paused, inactive, resume, inactive, resume"
as 
[inactive silently fulfilled]
FAIL: Unexpected lifecycle transition: paused

If you want to also fail on inactive, resume, inactive, resume, you should also set the expectations' [assertForOverFulfill](https://developer.apple.com/documentation/xctest/xctestexpectation/2806575-assertforoverfulfill?language=objc) to `YES`.  And if you expect the fulfillment to be more than one, you can also set [expectedFulfillmentCount](https://developer.apple.com/documentation/xctest/xctestexpectation/2806572-expectedfulfillmentcount?language=objc).

Reply:
I was halfway changing it to the way you said, but then I noticed that they both currently do the same thing. They both assert order, check on the same granularity and since NSNotificationCenter posts are synchronous, they also fail on unexpected events at the time of unexpected events. 

I think I still prefer the present way which at places lists out the total map of expected behavior.

Reply:
Here are the new tests after the fix https://github.com/xster/engine/compare/lifecycle-test...xster:fix-ios-lifecycle?expand=1